### ISSUE TITLE
Differences in `Base.Iterators.partition` and `IterTools.partition`

### ISSUE THREAD

**Issue by ssfrr** [JuliaCollections/IterTools.jl#67]
There are some differences between `Base.Iterators.partition` and `IterTools.partition`. I brought it up on Slack and @oxinabox asked me to file an issue.

1. `Base.Iterators.partition` returns each chunk as an `Array`, whereas `IterTools.partition` returns a `Tuple`
2. In the case that the last chunk is smaller than the requested partition size, `Base.Iterators.partition` returns a shorter chunk, whereas `IterTools.partition` drops the last values.
3. `IterTools.partition` accepts a `step` argument that determines the hop size for each chunk.

**Comment by ssfrr** [comment]
update - I think on `master` `Base.Iterators.partition` now returns views into the original array, not `Array`s.

This might (?) be the last inconsistency that would need to be addressed before something like #30 could be resolved. How do folks want to resolve the differences listed above? I'd propose using the `Base.Iterators` behavior for (1) and (2), but keeping the `step` argument.

One path forward would be to add the `step` argument as a PR to `Base`, then just reexport `Base.Iterators.partition`. This would also take advantage of the recent improvements that @mbauman made.

**Comment by omus** [comment]
Change to make `Base.Iterators.partition` return views was done in: https://github.com/JuliaLang/julia/pull/33533

**Comment by goretkin** [comment]
I'm in favor of keeping the current behavior of `IterTools.partition`, perhaps under a different name, to avoid confusion with `Base`. 

For example, implementing `Base.diff`, but for iterables instead of just AbstractVectors is very natural in terms of `IterTools.partition(it, 2, 1)`, and the fact that it produces a stream of tuples, which have a length known at compile time is a benefit.

**Comment by ssfrr** [comment]
Yeah, it's sort of a shame that the implementation needs to be different depending on whether it's designed for long or short partitions (which would use Arrays or Tuples, respectively). For instance, in my use case (streaming audio) I often want to chop the stream into chunks of 512 or 1024 elements, so Tuples aren't good for that. Maybe the partition type should be provided as an optional argument to `partition`?

**Comment by goretkin** [comment]
Oh, good. That's an important use too. `Base.Iterators.partition` returns views into the array, instead of copies, now. Does that fit the audio chunking need?

**Comment by ssfrr** [comment]
Yeah, array views are good for me
